Applications/Spire/Include/Spire/KeyBindings/KeyBindings.hpp

```diff
@@ -24,58 +24,127 @@ namespace Details {
+          using Type = typename Details::AddOptionalToVariant<
+            typename Details::AddOptionalToVariant<
+            typename Details::AddOptionalToVariant<
+            typename Details::AddOptionalToVariant<
+            typename Details::ToVariantOfOptionals<Nexus::Tag::Type>::Type,
+            Nexus::Side>::Type, Nexus::OrderType>::Type,
+            Nexus::Region>::Type, Nexus::TimeInForce>::Type;
+
+          //! Constructs a Tag.
+          /*!
+            \param key The key of the tag.
+            \param value The value of the tag.
+          */
+          template<typename T>
+          Tag(int key, T value);
+
+          //! Constructs a Tag.
+          /*!
+            \param key The key of the tag.
+            \param value The value of the tag.
+          */
+          template<typename T>
+          Tag(int key, boost::optional<T> value);
+
+          //! Returns the key of the tag.
+          int get_key() const;
+
+          //! Returns the value of the tag.
+          /*!
+            \tparam T The type of tag's value.
+            \return The value of the tag, boost::none if the type of the value
+                    is not T.
+          */
+          template<typename T>
+          boost::optional<T> get_value() const;
+
+          //! Returns the value of the tag.
+          /*!
+            \return The value of the tag contained in a variant of all possible
+                    types.
+          */
+          Type get_value() const;
+
+          //! Checks whether two tags are equal.
+          bool operator ==(const Tag& other) const;
+
+          //! Checks whether two tags are not equal.
+          bool operator !=(const Tag& other) const;
+
+        private:
+          friend std::ostream& operator <<(std::ostream& out,
+            const KeyBindings::Tag& tag);
+
+          int m_key;
           Type m_value;
-        };
-
-        //! The name of the action.
-        std::string m_name;
-
-        //! The type of the order.
-        boost::optional<Nexus::OrderType> m_type;
-
-        //! The side of the order.
-        boost::optional<Nexus::Side> m_side;
-
-        //! The expiry of the order.
-        boost::optional<Nexus::TimeInForce> m_time_in_force;
-
-        //! The quantity of the order.
-        boost::optional<Nexus::Quantity> m_quantity;
+      };
 
-        //! The list of custom tags.
-        std::vector<CustomTag> m_tags;
+      //! Stores a description of an order submission action.
+      class OrderAction {
+        public:
+
+          //! Constructs an unnamed OrderAction with no tags.
+          OrderAction() = default;
+
+          //! Constructs an OrderAction.
+          /*!
+            \param name The name of the action.
+            \param tags The list of the tags.
+          */
+          OrderAction(std::string name, std::vector<Tag> tags);
+
+          //! Returns the name of the action.
+          std::string get_name() const;
```

Comment:
Return ```const &```.

Reply:
I can, but as there are public setters for name and tags I thought it might be considered unsafe.
Should I also return `boost::optional<const Tag&>` from `get_tag` then?

Reply:
Yes, ```optional<T&>``` is the reason our codebase doesn't use ```std::optional```.